src/main/java/org/inferred/freebuilder/processor/excerpt/CheckedSetMultimap.java

```diff
@@ -0,0 +1,114 @@
+import com.google.common.collect.Multimaps;
+import com.google.common.collect.SetMultimap;
+
+import org.inferred.freebuilder.processor.util.Excerpt;
+import org.inferred.freebuilder.processor.util.ParameterizedType;
+import org.inferred.freebuilder.processor.util.SourceBuilder;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+/**
+ * Excerpts defining a multimap implementation that delegates to a provided put method to perform
+ * entry validation and insertion into a backing multimap.
+ */
+public class CheckedSetMultimap {
+
+  public static Set<Excerpt> excerpts() {
+    return ImmutableSet.<Excerpt>builder()
+        .addAll(CheckedSet.excerpts())
+        .add(CHECKED_SET_MULTIMAP)
+        .build();
+  }
+
+  private static final Excerpt CHECKED_SET_MULTIMAP = new Excerpt() {
+    @Override
+    public void addTo(SourceBuilder code) {
+      ParameterizedType biConsumer = code.feature(FUNCTION_PACKAGE).biConsumer().orNull();
+      if (biConsumer == null) {
+        return;
+      }
+      code.addLine("")
+          .addLine("/**")
+          .addLine(" * A multimap implementation that delegates to a provided put method")
+          .addLine(" * to perform entry validation and insertion into a backing multimap.")
+          .addLine(" */")
+          .addLine("private static class CheckedSetMultimap<K, V> extends %s<K, V> {",
+              ForwardingSetMultimap.class)
+          .addLine("")
+          .addLine("  private final %s<K, V> multimap;", SetMultimap.class)
+          .addLine("  private final %s<K, V> put;", biConsumer.getQualifiedName())
+          .addLine("")
+          .addLine("  CheckedSetMultimap(%s<K, V> multimap, %s<K, V> put) {",
+              SetMultimap.class, biConsumer.getQualifiedName())
+          .addLine("    this.multimap = multimap;")
+          .addLine("    this.put = put;")
+          .addLine("  }")
+          .addLine("")
+          .addLine("  @Override protected %s<K, V> delegate() {", SetMultimap.class)
+          .addLine("    return multimap;")
+          .addLine("  }")
+          .addLine("")
+          .addLine("  @Override public boolean put(@%1$s K key, @%1$s V value) {",
+              Nullable.class)
+          .addLine("    put.accept(key, value);")
+          .addLine("    return true;")
+          .addLine("  }")
+          .addLine("")
+          .addLine("  @Override public boolean putAll(@%s K key, %s<? extends V> values) {",
+              Nullable.class, Iterable.class)
+          .addLine("    boolean anyModified = false;")
+          .addLine("    for (V value : values) {")
+          .addLine("      put.accept(key, value);")
+          .addLine("      anyModified = true;")
+          .addLine("    }")
+          .addLine("    return anyModified;")
+          .addLine("  }")
+          .addLine("")
+          .addLine("  @Override public boolean putAll(%s<? extends K, ? extends V> multimap) {",
```

Comment:
is this just a hangover from Map? seems weird to return true if the input wasn't empty.